File: src/409.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    int longestPalindrome(string s) {
        int retVal = 0;

        unordered_map<char, int> frequency;
        for (char c : s) {
            ++frequency[c];
        }

        for (auto p : frequency) {
            int value = p.second;
            retVal += (value / 2 * 2);
            if ((value % 2 == 1) && (retVal % 2 == 0)) {
                ++retVal;
            }
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        string s;
    };
    vector<subject> testData{{"abccccdd"}, {"a"}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: s = "abccccdd"
     *  Output: 7
     *
     *  Input: s = "a"
     *  Output: 1
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: s = \"" << testData[i].s << "\"\n";

        answer = cSolution.longestPalindrome(testData[i].s);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/409.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def longestPalindrome(self, s: str) -> int:
        retVal = 0

        frequency = Counter(s)
        for value in frequency.values():
            retVal += (value // 2 * 2)
            if (retVal % 2 == 0) and (value % 2 == 1):
                retVal += 1

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for s in ["abccccdd", "a"]:
            # /* Example
            #  *  Input: s = "abccccdd"
            #  *  Output: 7
            #  *
            #  *  Input: s = "a"
            #  *  Output: 1
            #  */
            logging.info("Input: s = \"%s\"", s)

            retVal = pSolution.longestPalindrome(s)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
